pkg/services/store/entity/db/migrations/resource_mig.go

```diff
@@ -0,0 +1,151 @@
+		},
+		Indices: []*migrator.Index{
+			{Cols: []string{"group", "resource"}, Type: migrator.UniqueIndex},
+		},
+	})
+
+	tables = append(tables, migrator.Table{
+		Name: "resource", // write only log?  all events
+		Columns: []*migrator.Column{
+			// SnowflakeID -- Each Create/Update/Delete call is an event
+			// Using snowflake ID doubles this field as an approximate timestamp
+			{Name: "event", Type: migrator.DB_BigInt, Nullable: false, IsPrimaryKey: true},
+
+			// This will be null on insert, and then updated once we are ready to commit the transaction
+			{Name: "rv", Type: migrator.DB_BigInt, Nullable: true},
+			{Name: "previous_rv", Type: migrator.DB_BigInt, Nullable: true}, // needed?
+
+			// Allows fast search for the first page in any query.
+			// Subsequent pages must use MAX(rv) AND is_compacted=false GROUP ...
+			{Name: "is_current", Type: migrator.DB_Bool, Nullable: false},
+
+			// Indicates that this is no longer the current version
+			// This value is updated every few minutes and makes the paged queries more efficient
+			{Name: "is_compacted", Type: migrator.DB_Bool, Nullable: false},
+
+			// Properties that exist in path/key (and duplicated in the json value)
+			{Name: "group", Type: migrator.DB_NVarchar, Length: 190, Nullable: false},
+			{Name: "api_version", Type: migrator.DB_NVarchar, Length: 32, Nullable: false},
+			{Name: "namespace", Type: migrator.DB_NVarchar, Length: 63, Nullable: true}, // namespace is not required (cluster scope)
+			{Name: "resource", Type: migrator.DB_NVarchar, Length: 190, Nullable: false},
+			{Name: "name", Type: migrator.DB_NVarchar, Length: 190, Nullable: false},
+
+			// The operation that wrote this resource version
+			// 1: created, 2: updated, 3: deleted
+			{Name: "operation", Type: migrator.DB_Int, Nullable: false},
+
+			// Optional Commit message (currently only used for dashboards)
+			{Name: "message", Type: migrator.DB_Text, Nullable: false}, // defaults to empty string
+
+			// The k8s resource JSON text (without the resourceVersion populated)
+			{Name: "value", Type: migrator.DB_MediumText, Nullable: false},
+
+			// Content hash -- this is appropriate to use for an etag value
+			{Name: "hash", Type: migrator.DB_NVarchar, Length: 32, Nullable: false},
+
+			// Path to linked blob (or null).  This blob may be saved in SQL, or in an object store
+			{Name: "blob_path", Type: migrator.DB_NVarchar, Length: 1024, Nullable: true},
+		},
+		Indices: []*migrator.Index{
+			{Cols: []string{"rv"}, Type: migrator.UniqueIndex},
+			{Cols: []string{"is_current"}, Type: migrator.IndexType},
+			{Cols: []string{"is_compacted"}, Type: migrator.IndexType},
+			{Cols: []string{"operation"}, Type: migrator.IndexType},
+			{Cols: []string{"namespace"}, Type: migrator.IndexType},
+			{Cols: []string{"group", "resource", "name"}, Type: migrator.IndexType},
+			{Cols: []string{"blob_path"}, Type: migrator.IndexType},
```

Comment:
For MySQL at least, it'll choose a single index for each query, meaning that we'll need is_current to be part of the other indexes to avoid having to search through all historical rows.

https://dev.mysql.com/doc/refman/8.4/en/mysql-indexes.html

Postgres is a bit smarter and will be able to combine indexes to optimize queries, but will still have to build the bitmaps and AND them together to find the rows we want.

https://www.postgresql.org/docs/current/indexes-bitmap-scans.html

This is one of the reasons we have the entity table as well as entity_history, since the vast majority of queries can be answered without needing to read historical revisions.